keps/sig-network/1880-multiple-service-cidrs/README.md

```diff
@@ -0,0 +1,954 @@
+
+During this time, there is a chance that an apiserver tries to allocate this IPAddress, with a possible situation where
+2 Services has the same IPAddress. In order to avoid it, the Allocator will not delete an IP from its local cache
+until it verifies that the consumer associated to that IP has been deleted too.
+
+- Valid IPAddress and ServiceCIDRConfig but no Service
+
+The IPAddress will be deleted and an event generated if the controller determines that the IPAddress
+is orphan [(see Allocator section)](#Allocator)
+
+- IPAddress referencing recreated Object (different UID)
+
+1. User created Gateway "foo"
+2. Gateway controller allocated IP and ref -> "foo"
+3. User deleted gateway "foo"
+4. Gateway controller doesn't delete the IP (leave it for GC)
+5. User creates a new Gateway "foo"
+6. apiserver repair loop finds the IP with a valid ref to "foo"
+
+If the new gateway is created before the apiserver observes the delete, apiserver will find that gateway "foo"
+still exists and can't release the IP. It can't peek inside "foo" to see if that is the right IP because it is
+a type it does not know. If it knew the UID it could see that "foo" UID was different and release the IP.
+The IPAddress will use the UID to reference the parent to avoid problems in this scenario.
+
+#### Resizing Service IP Ranges
+
+One of the most common problems users may have is how to scale up or scale down their Service CIDRs.
+
+Let's see an example on how these operations will be implemented.
+
+Assume we have configured a Service CIDR 10.0.0.0/24 and its fully used, we can:
+
+1. Add another /24 Service CIDR 10.0.1.0/24 and keep the previous one
+2. Add an overlapping larger Service CIDR 10.0.0.0/23
+
+After 2., the user can now remove the first /24 Service CIDR, since the new Service CIDR covers all the existing IP Addresses
+
+The same applies for scaling down, meanwhile the new Service CIDR contains all the existing IPAddresses, the old
+Service CIDR will be safely deleted.
+
+However, there can be a race condition during the operation of scaling down, since the propagation of the
+deletion can take some time, one allocator can successfully allocate an IP address out of new Service CIDR (but still
+inside of the old Service CIDR).
+
+There is one controller that will periodically check that the 1-on-1 relation between IPAddresses and Services is
+correct, and will start sending events to warn the user that it has to fix/recreate the corresponding Service.
+
+#### API
+
+```go
+// ServiceCIDRConfig defines a range of IPs using CIDR format (192.168.0.0/24 or 2001:db2::0/64).
+type ServiceCIDRConfig struct {
+ metav1.TypeMeta   `json:",inline"`
+ metav1.ObjectMeta `json:"metadata,omitempty"`
+
+ Spec   ServiceCIDRConfigSpec   `json:"spec,omitempty"`
+}
+
+
+// ServiceCIDRConfigSpec describe how the ServiceCIDRConfig's specification looks like.
+type ServiceCIDRConfigSpec struct {
```

Comment:
Opportunity for edge case:

* Start a single-stack apiserver with a v4 range
* It creates a ServiceCIDRConfig with v4
* Change config to dual-stack and restart
* It does not find an exact match
* It creates a ServiceCIDRConfig with v4 and v6
* Now we have 2 ServiceCIDRConfig forever

Maybe the above startup pseudocode changes:

```
  if multiple ranges match exactly my flags
   sort them by creation time
   log
   delete all but one
  endif
  if multiple ranges match exactly my flags (or a single-family subset of)
    log
    delete all subsets, leaving the largest set that exactly matches on at least on family
```

Reply:
yeah, we should be able to support upgrades and deal with things like https://github.com/kubernetes/enhancements/pull/3365/files#r902005526

Reply:
updated the bootsrap with the "merge" instead of "delete"